keps/sig-instrumentation/2831-kubelet-tracing/README.md

```diff
@@ -0,0 +1,364 @@
+Trace data is structured, providing the detail necessary to debug requests across service boundaries.
+As more core components are instrumented, Kubernetes becomes easier to monitor, manage, and troubleshoot.
+
+### Definitions
+
+**Span**: The smallest unit of a trace.  It has a start and end time. Spans are the building blocks of a trace.    
+**Trace**: A collection of Spans which represents work being done by a system. A record of the path of requests through a system.    
+**Trace Context**: A reference to a Trace that is designed to be propagated across component boundaries.    
+
+### Goals
+
+* The kubelet generates and exports spans for reconcile loops it initiates and for outgoing requests to the CRI. An example of a reconcile loop is the
+creation of a pod. Pod creation involves pulling the image, creating the pod sandbox and creating the container. Those would be linked together into a
+single trace to diagnose latency issues within that flow.
+
+### Non-Goals
+
+* Tracing in other kubernetes controllers besides the kubelet
+* Replace existing logging, metrics
+* Change metrics or logging (e.g. to support trace-metric correlation)
+* Access control to tracing backends
+* Add tracing to components outside kubernetes (e.g. etcd client library).
+
+## Proposal
+
+### User Stories
+
+Since this feature is for diagnosing problems with the kubelet, it is targeted at cluster operators, administrators, and cloud vendors that 
+manage kubernetes nodes and core services.
+
+For the following use-cases, I can deploy an OpenTelemetry collector agent as a DaemonSet to collect kubelet trace data from each node's host network.
+From there, OpenTelemetry trace data can be exported to a tracing backend of choice.  I can use the kubelet's `--opentelemetry-config-file flag` with
+the default URL to make the kubelet send its spans to the collector. Alternatively, I can point the kubelet at an OpenTelemetry collector listening on
+a different port or URL if I need to.
+
+#### Continuous Trace Collection
+
+As a cluster administrator or cloud provider, I would like to collect gRPC and HTTP trace data from the transactions between the API server and the 
+kubelet and interactions with a node's container runtime (Container Runtime Interface) to debug cluster problems.  I can set the `SamplingRatePerMillion`
+in the configuration file to a non-zero number to have spans collected for a small fraction of requests. Depending on the symptoms I need to
+debug, I can search span metadata or specific nodes to find a trace which displays the symptoms I am looking to debug.
+The sampling rate for trace exports can be configured based on my needs. I can collect each node's kubelet trace data as distinct tracing services
+to diagnose node issues.
+
+##### Example Scenarios
+
+* Latency or timeout experienced when:
+    * Attach or exec to running containers
```

Comment:
This will make a neat demo!

Reply:
definitely! I have a setup that is easily reproducible (I believe I used your demos to put it all together, too) 
https://github.com/sallyom/otel-kubeadm